File: leetcode/78.search-a-2d-matrix-ii/suhy-binary.js
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
const binary = (matrix, target) => {
    const searchHelper = (left, top, right, bottom) => {
        if (left > right || top > bottom) return false;
        if (target < matrix[top][left] || target > matrix[bottom][right]) return false;
        
        const mid = left + Math.floor((right - left) / 2);
        let row = top;
        while (bottom >= row && matrix[row][mid] <= target) {
            if (matrix[row][mid] == target) {
                return true;
            }
            row += 1
        }
        return searchHelper(left, row, mid-1, bottom) || searchHelper(mid+1, top, right, row-1)
    }
    const searchMatrix = () => {
        if (matrix.length === 0) return false;
        return searchHelper(0, 0, matrix[0].length-1, matrix.length-1)
    }
    return {
        searchMatrix
    };
};
const searchMatrix = (matrix, target) => {
    return binary(matrix, target).searchMatrix();    
};

File: leetcode/78.search-a-2d-matrix-ii/suhy-binary.rb
# @param {Integer[][]} matrix
# @param {Integer} target
# @return {Boolean}
class Solution
    def initialize(matrix, target)
        @matrix = matrix
        @target = target
    end
    
    def searchRec( left, up, right, down )
        return false if left > right || up > down 
        return false if @target < @matrix[up][left] || @target > @matrix[down][right]
        
        mid = left + (right - left) / 2
        row = up
        while row <= down && @matrix[row][mid] <= @target
            return true if @matrix[row][mid] == @target
            
            row += 1
        end
        searchRec(left, row, mid-1, down) || searchRec(mid+1, up, right, row-1)
    end
    
    def search_matrix()
        return false if @matrix.nil? || @matrix.length == 0
        
        return searchRec(0, 0, @matrix[0].length-1, @matrix.length-1)
    end
end

def search_matrix(matrix, target)
    Solution.new(matrix, target).search_matrix
end
